internal/build/signify.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// signFile reads the contents of an input file and signs it (in armored format)
+// with the key provided, placing the signature into the output file.
+
+package build
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"crypto/ed25519"
+)
+
+var (
+	errInvalidKeyHeader = errors.New("Incorrect key header")
+	errInvalidKeyLength = errors.New("invalid, key length != 42")
+)
+
+func readSKey(key []byte) (ed25519.PrivateKey, error) {
+	if len(key) != 104 {
+		return nil, errInvalidKeyLength
+	}
+
+	if string(key[:2]) != "Ed" {
+		return nil, errInvalidKeyHeader
+	}
+
+	return ed25519.PrivateKey(key[40:]), nil
+
+}
+
+// SignifySignFile creates a signature of the input file.
+func SignifySignFile(input string, output string, key string) error {
+	in, err := os.Open(input)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+
+	out, err := os.Create(output)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	keydata, err := base64.StdEncoding.DecodeString(key)
+	if err != nil {
+		return err
+	}
+	skey, err := readSKey(keydata)
+	if err != nil {
+		return err
+	}
+
+	filedata, err := ioutil.ReadAll(in)
+	if err != nil {
+		return err
+	}
+
+	sigdata := []byte("Ed")
+	copy(sigdata, keydata[:2])
+	sigdata = append(sigdata, keydata[32:40]...)
+	sigdata = append(sigdata, ed25519.Sign(skey, filedata)...)
```

Comment:
Wow, this looks really funky! At a glance, it looks like you're just happily appending the private key to the `sigdata`, which is then output in the message. 
Not until one checks into `readSKey`, it becomes apparent that the private part after the first `40` bytes, which is not included here. 

Perhaps you can make this more obvious, like 
```
privKey := keydata[40:]
pubKey := keydata[32:40]
...
sigdata = append(sigdata, pubKey...)
```
Also, I don't quite understand -- what is sitting on bytes `2-32`, which is totally ignored?

Reply:
2-32 it contains data that is indicative of how the key got generated and also a checksu, which is not to be published in the signature.